fainder/execution/parallel_processing.py

```diff
@@ -0,0 +1,247 @@
+        hists: list[tuple[int | np.integer[Any], Histogram]] = load_input(
+            histogram_path, "histograms"
+        )
+        if hists is None:
+            logger.error(f"Worker {worker_id} failed to load histograms from {histogram_path}")
+            continue
+
+        # Merge into worker's histogram dictionary
+        for id_, hist in hists:
+            _worker_state.hists[id_] = hist
+
+    logger.info(
+        f"Worker {worker_id} initialized with {len(_worker_state.hists)}"
+        f" histograms from {len(histogram_paths)} chunks"
+    )
+
+
+def process_hist_chunk(
+    query: PercentileQuery, id_filter: NDArray[np.uint32]
+) -> NDArray[np.uint32]:
+    """Process the chunk of histograms assigned to this worker."""
+    global _worker_state  # noqa: PLW0602
+    from fainder.execution.percentile_queries import query_histogram
+
+    if _worker_state.hists is None:
+        logger.error("Worker called without being initialized!")
+        return np.array([], dtype=np.uint32)
+
+    # Filter histograms by ID
+    filtered_hists: list[tuple[int | np.integer[Any], Histogram]] = [
+        (id_f, _worker_state.hists[id_f]) for id_f in id_filter if id_f in _worker_state.hists
+    ]
+
+    # Process the histograms
+    return np.fromiter(
+        (
+            np.uint32(id_)
+            for id_, hist in filtered_hists
+            if query_histogram(hist, estimation_mode="over", query=query, density=True)
+        ),
+        dtype=np.uint32,
+    )
+
+
+def partition_histogram_ids(
+    hist_ids: list[int],
+    num_partitions: int,
+    chunk_layout: FainderChunkLayout = FainderChunkLayout.CONTIGUOUS,
+) -> dict[int, set[int]]:
+    """Partition histogram IDs into roughly equal chunks.
+
+    Args:
+        hist_ids: List of histogram IDs to partition
+        num_partitions: Number of partitions to create
+        contiguous: If True, use contiguous chunks; if False, distribute in round-robin fashion
+
+    Returns:
+        Dictionary mapping partition ID to list of histogram IDs
+    """
+    chunks: dict[int, set[int]] = {i: set() for i in range(num_partitions)}
+
+    if chunk_layout == FainderChunkLayout.CONTIGUOUS:
+        # Original contiguous chunking strategy
+        chunk_size = len(hist_ids) // num_partitions
+        remainder = len(hist_ids) % num_partitions
+
+        start_idx = 0
+        for i in range(num_partitions):
+            end_idx = start_idx + chunk_size + (1 if i < remainder else 0)
+            chunks[i] = set(hist_ids[start_idx:end_idx])
+            start_idx = end_idx
+    elif chunk_layout == FainderChunkLayout.ROUND_ROBIN:
+        # Round-robin distribution for more balanced workload
+        current_partition = 0
+        for hist_id in hist_ids:
+            chunks[current_partition].add(hist_id)
+            if current_partition == num_partitions - 1:
+                current_partition = 0
+            else:
+                current_partition += 1
```

Comment:
If you want to play code golf, you could rewrite this to 
`current_partition = (current_partition + 1) % num_partitions`